## A list of useful software resources:

**[Plink:](http://zzz.bwh.harvard.edu/plink/)** a free, open-source whole genome association analysis toolset, designed to perform a range of basic, large-scale analyses in a computationally efficient manner.

**[The Galaxy Project:](https://galaxyproject.org/)** a scientific workflow, data integration, and data and analysis persistence and publishing platform that aims to make computational biology accessible to research scientists that do not have computer programming or systems administration experience. Although it was initially developed for genomics research, it is largely domain agnostic and is now used as a general bioinformatics workflow management system.


**[BioPython:](http://biopython.org/)** an open-source collection of non-commercial Python tools for computational biology and bioinformatics, created by an international association of developers. It contains classes to represent biological sequences and sequence annotations, and it is able to read and write to a variety of file formats. It also allows for a programmatic means of accessing online databases of biological information, such as those at NCBI. Separate modules extend Biopython's capabilities to sequence alignment, protein structure, population genetics, phylogenetics, sequence motifs, and machine learning

**[BioConductor:](https://www.bioconductor.org/)** a free, open source and open development software project for the analysis and comprehension of genomic data generated by wet lab experiments in molecular biology. Bioconductor is based primarily on the statistical R programming language, but does contain contributions in other programming languages. It has two releases each year that follow the semiannual releases of R. At any one time there is a release version, which corresponds to the released version of R, and a development version, which corresponds to the development version of R.

**[LDpred:](https://github.com/bvilhjal/ldpred)** a method that infers the posterior mean effect size of each marker by using a prior on effect sizes and LD information from an external reference panel and calculates PGS.


 **[PRSice:](http://prsice.info/)** a software package for calculating, applying, evaluating and plotting the results of polygenic risk scores.
